File: zig/perfect-numbers/perfect_numbers.zig
pub const Classification = enum {
    deficient,
    perfect,
    abundant,
};

/// Asserts that `n` is nonzero.
pub fn classify(n: u64) Classification {
    const half: u64 = n / 2;
    var sum: u64 = 0;
    for (1..half + 1) |num| {
        if (n % num == 0) {
            sum += num;
        }
    }

    // if (sum == n) {
    //     return Classification.perfect;
    // }
    //
    // return if (sum > n) Classification.abundant else Classification.deficient;

    if (n == sum) {
        return Classification.perfect;
    }

    return if (n > sum) Classification.deficient else Classification.abundant;
}

File: elixir/perfect-numbers/lib/perfect_numbers.ex
defmodule PerfectNumbers do
  @doc """
  Determine the aliquot sum of the given `number`, by summing all the factors
  of `number`, aside from `number` itself.

  Based on this sum, classify the number as:

  :perfect if the aliquot sum is equal to `number`
  :abundant if the aliquot sum is greater than `number`
  :deficient if the aliquot sum is less than `number`
  """
  @spec classify(number :: integer) :: {:ok, atom} | {:error, String.t()}
  def classify(number) when number < 1,
    do: {:error, "Classification is only possible for natural numbers."}

  def classify(1), do: {:ok, :deficient}
  def classify(2), do: {:ok, :deficient}
  def classify(number), do: sum_factors(number)

  @spec sum_factors(n :: integer) :: {:ok, atom}
  defp sum_factors(n) do
    half = div(n, 2)
    sum = 2..half |> Enum.reduce(1, fn x, acc -> if rem(n, x) == 0, do: acc + x, else: acc end)

    cond do
      sum == n -> {:ok, :perfect}
      true -> if sum > n, do: {:ok, :abundant}, else: {:ok, :deficient}
    end
  end
end
